typing-c/src/C/Type.hs

```diff
@@ -0,0 +1,294 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module C.Type
+  ( -- * C types
+    Type(..)
+  , ArithmeticType(..)
+  , IntegralType(..)
+  , CharLikeType(..)
+  , IntLikeType(..)
+  , Sign(..)
+  , FloatLikeType(..)
+  , IntegerConversionRank(..)
+  , intLikeTypeSign
+  , charLikeTypeSizeInBits
+  , intLikeTypeSizeInBits
+  , intLikeTypeConversionRank
+  , showTypeAsCType
+
+  -- * Platform
+  , Platform(..), WordWidth(..), OS(..)
+  , hostPlatform
+
+  -- * Listing types
+  , enumerateTypeTuples
+  ) where
+
+-- base
+import Control.Arrow
+  ( second )
+import Data.Functor
+  ( (<&>) )
+import Data.Semigroup
+  ( Arg(..) )
+import Foreign.Storable
+  ( sizeOf )
+import System.Info qualified
+
+-- fin
+import qualified Data.Type.Nat as Fin
+
+-- vec
+import Data.Vec.Lazy ( Vec(..) )
+import Data.Vec.Lazy qualified as Vec
+  ( snoc )
+
+--------------------------------------------------------------------------------
+
+data WordWidth = WordWidth32 | WordWidth64
+  deriving stock ( Eq, Ord, Show )
+
+wordWidthInBits :: WordWidth -> Word
+wordWidthInBits = \case
+  WordWidth32 -> 32
+  WordWidth64 -> 64
+
+data OS = Windows | Posix
+  deriving stock ( Eq, Ord, Show )
+
+data Platform = Platform { platformWordWidth :: !WordWidth
+                         , platformOS        :: !OS }
+
+
+hostPlatform :: Platform
+hostPlatform =
+  Platform
+    { platformWordWidth =
+        case sizeOf @Int undefined of
+          4 -> WordWidth32
+          8 -> WordWidth64
+          w -> error $ "hostPlatform, unsupported word width " ++ show (8 * w)
+    , platformOS =
+        case System.Info.os of
+          "mingw32" -> Windows
+          _         -> Posix
+    }
+
+data Type
```

Comment:
```hs
data Type a
  = Void
  | Arithmetic !ArithmeticType
  | Ptr        !a
```

so that we can use e.g. `Type (Maybe Int)` for generating type family equations.